Ports /vg/'s SQL migration controller.

ShiftyRail opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request

When porting BSQL to /vg/, I noticed TG maintainers need to update the database manually for each PR pertaining to SQL.

On /vg/, we have a system which automatically updates the DB whenever a PR a merged if the contributor coded a migration file, so I figured I would port it to TG now that I got it to work for BSQL.

Testing done :
- download and compile TG's source code
- add the migration controller
- set up a local DB with WAMP and use the SQL schema to make it work
- create a very basic migration file (adding an empty table)
- check if the DB is updated in phpMyAdmin, look for roundstart runtimes, see if I can join. 

I've included the DB Major/Minor table in this PR, however do note that the migration controller essentially makes it redundant, as it has its own check to see if the database is up to date or not.

## Why It's Good For The Game

Let contributors modify the database more easily. Any maintainer can now merge a PR pertaining to SQL, instead of only the maintainer with DB access.

ShiftyRail committed:
Ports /vg/'s migration controller.

bobbahbrown commented on code/modules/migrations/migrations_controller.dm:
```diff
@@ -0,0 +1,114 @@
+
+/datum/migration_controller/New()
+	if(!setup())
+		world.log << "\[Migrations] ([id]): Setup() returned false, will not run migrations for this DBMS."
```
consider using log_sql or log_world

```suggestion
		log_sql("\[Migrations] ([id]): Setup() returned false, will not run migrations for this DBMS.")
```

bobbahbrown commented on code/modules/migrations/migrations_controller.dm:
```diff
@@ -0,0 +1,114 @@
+		world.log << "\[Migrations] ([id]): Setup() returned false, will not run migrations for this DBMS."
+	else
+		if(!hasTable(TABLE_NAME))
+			world.log << "\[Migrations] ([id]): Creating [format_table_name(TABLE_NAME)]"
```
consider using log_sql or log_world

```suggestion
			log_sql("\[Migrations] ([id]): Creating [format_table_name(TABLE_NAME)]")
```

bobbahbrown commented on code/modules/migrations/migrations_controller.dm:
```diff
@@ -0,0 +1,114 @@
+			var/list/pack[prepack.len]
+			for(var/datum/migration/M in newpacks[pkgID])
+				pack[M.id] = M
+				//world.log << "\[Migrations] [pkgID]#[M.id] = [M.type] - [M.name]"
```
dont leave comments in production code
```suggestion
```

bobbahbrown commented on code/modules/migrations/migrations_controller.dm:
```diff
@@ -0,0 +1,114 @@
+				//world.log << "\[Migrations] [pkgID]#[M.id] = [M.type] - [M.name]"
+			packages[pkgID]=pack
+			world.log << "\[Migrations] Loaded [pack.len] [id] DB migrations from package [pkgID]."
+		//VersionCheck()
```
dont leave commented out code in production code
```suggestion
```

bobbahbrown commented on code/modules/migrations/migrations_controller.dm:
```diff
@@ -0,0 +1,114 @@
+				pack[M.id] = M
+				//world.log << "\[Migrations] [pkgID]#[M.id] = [M.type] - [M.name]"
+			packages[pkgID]=pack
+			world.log << "\[Migrations] Loaded [pack.len] [id] DB migrations from package [pkgID]."
```
consider using log_sql or log_world

```suggestion
			log_sql("\[Migrations] Loaded [pack.len] [id] DB migrations from package [pkgID].")
```

bobbahbrown commented on code/modules/migrations/migrations_controller.dm:
```diff
@@ -0,0 +1,114 @@
+		UpdateAll()
+
+/datum/migration_controller/proc/getCurrentVersion(var/pkgID)
+	if(pkgID in db_states)
+		return db_states[pkgID]
+	else
+		return 0
```
```suggestion
	return (pkgID in db_states) && db_states[pkgID]
```

bobbahbrown commented on code/modules/migrations/migrations_controller.dm:
```diff
@@ -0,0 +1,114 @@
+/datum/migration_controller
```
[Please document your code to meet dmdoc requirements.](https://github.com/tgstation/tgstation/blob/master/.github/AUTODOC_GUIDE.md)

bobbahbrown commented on code/modules/migrations/migration.dm:
```diff
@@ -0,0 +1,40 @@
+ * @author Rob "N3X15" Nelson
+ */
+
+/datum/migration
```
[Please document your code to meet dmdoc requirements.](https://github.com/tgstation/tgstation/blob/master/.github/AUTODOC_GUIDE.md)

bobbahbrown commented on code/modules/migrations/migrations_controller.dm:
```diff
@@ -0,0 +1,114 @@
+		for(var/datum/migration/M in packages[pkgID])
+			if(M.id > latestVersionAvail)
+				latestVersionAvail = M.id
+		//world.log << "\[Migrations] Package [pkgID]: Current: [currentVersion], Avail: [latestVersionAvail]"
```
dont leave commented out code in production code
```suggestion
```

bobbahbrown commented on code/modules/migrations/migrations_controller.dm:
```diff
@@ -0,0 +1,114 @@
+				latestVersionAvail = M.id
+		//world.log << "\[Migrations] Package [pkgID]: Current: [currentVersion], Avail: [latestVersionAvail]"
+		if(latestVersionAvail > currentVersion)
+			world.log << "\[Migrations] *** [pkgID] is behind [latestVersionAvail-currentVersion] versions!"
```
consider using log_sql or log_world

```suggestion
			log_sql("\[Migrations] *** [pkgID] is behind [latestVersionAvail-currentVersion] versions!")
```

bobbahbrown commented on code/modules/migrations/migrations_controller.dm:
```diff
@@ -0,0 +1,114 @@
+			if(M.id > to_version)
+				to_version = M.id
+	if(from_version == to_version)
+		world.log << "\[Migrations] [pkgID] is up to date."
```
consider using log_sql or log_world

```suggestion
		log_sql("\[Migrations] [pkgID] is up to date.")
```

bobbahbrown commented on code/modules/migrations/migrations_controller.dm:
```diff
@@ -0,0 +1,114 @@
+	if(from_version == to_version)
+		world.log << "\[Migrations] [pkgID] is up to date."
+		return
+	world.log << "\[Migrations] Updating [pkgID] from [from_version] to [to_version]..."
```
consider using log_sql or log_world

```suggestion
	log_sql("\[Migrations] Updating [pkgID] from [from_version] to [to_version]...")
```

bobbahbrown commented on code/modules/migrations/migrations_controller.dm:
```diff
@@ -0,0 +1,114 @@
+	for(var/datum/migration/M in package)
+		if(M.id > from_version && M.id <= to_version)
+			if(!M.up())
+//				to_chat(world, "[log_text] <span style='font-weight:bold;color:red;'>FAIL</span><br>Failed to process migration [pkgID] #[M.id]!")
```
dont leave commented out code in production code

```suggestion
```

bobbahbrown commented on code/modules/migrations/migrations_controller.dm:
```diff
@@ -0,0 +1,114 @@
+		if(M.id > from_version && M.id <= to_version)
+			if(!M.up())
+//				to_chat(world, "[log_text] <span style='font-weight:bold;color:red;'>FAIL</span><br>Failed to process migration [pkgID] #[M.id]!")
+				world.log << "Failed to process migration [pkgID] #[M.id]"
```
consider using log_sql or log_world

```suggestion
				log_sql("Failed to process migration [pkgID] #[M.id]")
```

bobbahbrown commented on code/modules/migrations/migrations_controller.dm:
```diff
@@ -0,0 +1,114 @@
+				return FALSE
+			else
+				M.execute("REPLACE INTO [format_table_name(TABLE_NAME)] (pkgID,version) VALUES ('[pkgID]',[M.id])") // SQLite also supports REPLACE.
+				world.log << "\[Migrations] Successfully applied [pkgID]#[M.id] ([M.name])"
```
consider using log_sql or log_world

```suggestion
				log_sql("\[Migrations] Successfully applied [pkgID]#[M.id] ([M.name])")
```

bobbahbrown commented on code/modules/migrations/migrations_controller.dm:
```diff
@@ -0,0 +1,114 @@
+			else
+				M.execute("REPLACE INTO [format_table_name(TABLE_NAME)] (pkgID,version) VALUES ('[pkgID]',[M.id])") // SQLite also supports REPLACE.
+				world.log << "\[Migrations] Successfully applied [pkgID]#[M.id] ([M.name])"
+	world.log << "\[Migrations] Done!"
```
consider using log_sql or log_world

```suggestion
	log_sql("\[Migrations] Done!")
```

bobbahbrown commented on code/modules/migrations/mysql/mysql_controller.dm:
```diff
@@ -0,0 +1,57 @@
+		warning("Something wrong with SSdbcore.NewQuery()")
+		return FALSE
+	qdel(Q)
+	//testing("MySQL is okay")
```
dont leave commented out code in production code

```suggestion
```

bobbahbrown commented on code/modules/migrations/mysql/mysql_controller.dm:
```diff
@@ -0,0 +1,57 @@
+GLOBAL_DATUM(migration_controller_mysql, /datum/migration_controller/mysql)
+
+/datum/migration_controller/mysql
```
[Please document your code to meet dmdoc requirements.](https://github.com/tgstation/tgstation/blob/master/.github/AUTODOC_GUIDE.md)

bobbahbrown commented on code/modules/migrations/mysql/mysql_migration.dm:
```diff
@@ -0,0 +1,63 @@
+/datum/migration/mysql
```
[Please document your code to meet dmdoc requirements.](https://github.com/tgstation/tgstation/blob/master/.github/AUTODOC_GUIDE.md)

bobbahbrown commented on code/modules/migrations/mysql/mysql_migration.dm:
```diff
@@ -0,0 +1,63 @@
+/datum/migration/mysql/query(var/sql)
+	var/datum/DBQuery/query = db.NewQuery(sql)
+	if(!query.Execute())
+		world.log << "Error in [package]#[id]: [query.ErrorMsg()]"
```
consider using log_sql or log_world

```suggestion
		log_sql("Error in [package]#[id]: [query.ErrorMsg()]")
```

bobbahbrown commented on code/modules/migrations/mysql/mysql_migration.dm:
```diff
@@ -0,0 +1,63 @@
+/datum/migration/mysql/hasResult(var/sql)
+	var/datum/DBQuery/query = db.NewQuery(sql)
+	if(!query.Execute())
+		world.log << "Error in [package]#[id]: [query.ErrorMsg()]"
```
consider using log_sql or log_world

```suggestion
		log_sql("Error in [package]#[id]: [query.ErrorMsg()]")
```

bobbahbrown commented on code/modules/migrations/mysql/mysql_migration.dm:
```diff
@@ -0,0 +1,63 @@
+/datum/migration/mysql/execute(var/sql)
+	var/datum/DBQuery/query = db.NewQuery(sql)
+	if(!query.Execute())
+		world.log << "Error in [package]#[id]: [query.ErrorMsg()]"
```
consider using log_sql or log_world

```suggestion
		log_sql("Error in [package]#[id]: [query.ErrorMsg()]")
```

bobbahbrown commented on code/modules/migrations/ss13/0xx-example.dm:
```diff
@@ -0,0 +1,31 @@
+
```
probably better to just document this somewhere instead of including the mock file

consider our hack.md

bobbahbrown commented on code/modules/migrations/migrations_controller.dm:
```diff
@@ -0,0 +1,114 @@
+		var/list/newpacks[0]
+		for(var/mtype in typesof(/datum/migration)-list(/datum/migration))
+			var/datum/migration/M = new mtype(src)
+			if(M.package == "" || M.name == "" || M.dbms != id)
```
reduce errors in the future

ps. not entirely sure this will operate correctly, I'm pretty sure but not 100% so just check that initial won't grab the initial of the child's type instead of the parent (datum/migration) -- it would be unexpected but I don't know for sure

```suggestion
			if(M.package == initial(M.package) || M.name == initial(M.name) || M.dbms != id)
```

bobbahbrown commented:
By the way this absolutely needs a changelog, please use the PR format correctly

Cyberboss commented:
The problem with us and migrations is we have multiple servers all using one DB. How does this account for that

Cyberboss commented:
@MrStonedOne @optimumtact @Jordie0608

Cyberboss commented:
This is honestly something I feel like would be good for us if we could make it work.

ShiftyRail commented:
I'm sorry about all the little logging and doc details, it's not something I'm used to check on my repo, same for changelog, we don't make :cl: for backend changes, but when in Rome, I'll do as Romans do.
As far as testing go, it *does* work on my machine. I can provide proof of that.

@Cyberboss Multiple servers shouldn't be a problem. Each migration is ran exactly once, updates the package number, and it's never ran again unless you roll it.

Cyberboss commented:
I'm more concerned about the case where what we consider a major update happens during a round when not all servers have the code yet (i.e. potential admin or ban table refactor). Though I'm not even sure how we currently deal with it.

Jordie0608 commented:
Given it's made for VG it seems to assume only one server so any PR needing schema and code update syncing, which is most of them, would need to have maintainers ensuring that all servers get updated and restart at the same time (or contributors blindly configuring database views). And then I guess just hoping there aren't any unnoticed issues that pop up once the game starts since a revert would need someone with DB access or another migration that rolls back the first.

This would also mean allowing the game server db accounts to have DDL permissions. Which imo is fairly unsafe.

That aside I can't test it yet but pretty sure you can use the package id for sql injection.

ShiftyRail commented:
You are correct in the sense that if you **remove** a column, a table, or something similar while multiple server are running, one of them will do the migration and update the DB for all of them, while other servers will have outdated code which assumes  said column exists, you will have runtimes and SQL errors.
However, this issue exists independently of a migration controller. 

>That aside I can't test it yet but pretty sure you can use the package id for sql injection.

I will use those protections your server has against VV and AdminProcCalls on migration datums, then.

>And then I guess just hoping there aren't any unnoticed issues that pop up once the game starts since a revert would need someone with DB access or another migration that rolls back the first.

The SQL code to undo a migration is contained in its `down()` proc.

ShiftyRail committed:
Logging and other stuff.

Co-authored-by: Bobbahbrown <bobbahbrown@gmail.com>

ShiftyRail committed:
Update code/modules/migrations/mysql/mysql_migration.dm

Co-authored-by: Bobbahbrown <bobbahbrown@gmail.com>

ShiftyRail commented:
Also, I realised that downstreams could be very interested in this as they can define their own migration packages who run independently of the ones TG has.

MrStonedOne commented:
The main blocker on something like this at /tg/ is that the game servers don't have DDL permissions so they literally couldn't run these migrations.


My solution to this issue is likely going to be to reduce the need for migrations. consistency and type enforcement is valid for certain parts of the database but like 90% could just be a json store and nothing would effectively change.

ShiftyRail commented:
Well then, the code is here and working. Might as well finish the doc while I'm at it.

Cyberboss commented:
Worst case, we could lock this behind a config so people can migrate their own databases using game code instead of [shittily thought out hacks](https://github.com/Cyberboss/tgstation/blob/TiredOfLinuxesShit/tools/LinuxOneShot/SetupProgram/PreCompile.sh#L97).

ShiftyRail committed:
Autodoc.

ShiftyRail committed:
Merge branch 'MigrationController' of https://github.com/ShiftyRail/tgstation into MigrationController

ShiftyRail commented:
I removed the example file (even though I personally think it's the best way to understand the thing) and added autodoc. I think. 

I'm not sure what I should put in the :cl:, to be honest. 

:cl: ShiftyRail, N3X15, /vg/station
code: Added a migration controller for easy database updates
/:cl:

Like this ? To me it's just going to confuse players if this is merged, but alright.

optimumtact commented:
this doesn't need a changelog

bobbahbrown commented:
@optimumtact

That is fair, apologies.

optimumtact commented:
I would hope server administrators of downstreams understand this change hwen they merge it in, but perhaps we need an operators log

Jordie0608 commented:
Schema change syncing isn't limited to removing parts of the schema; depends on the change but it can happen with many uses of alter i.e. changing a column from null to not null, adding a column, changing a datatype.

>The SQL code to undo a migration is contained in its down() proc.
I meant for like if a schema change is applied successfully, but then an issue with the PR is discovered post-merge and it needs to be reverted. The `down()` proc isn't actually called anywhere so you'd need to either make another migration to execute that or have someone with DB access around.

ShiftyRail commented:
I think I've added the doc. Tell me if I need to make it a config option or not.
The issue of DDL rights isn't really something I can do anything about.

The issue of multiple servers relying on a single DB isn't something addressed by this controller, but again, it is something that exists independently of it. But unless I'm missing something, this issue only exists if you somehow change the DB in a way that outdated code doesn't execute any more, such as changing the name of a table or the organisation of said table. 
If your change is *adding* a new table, or something which doesn't disturb the existing structure, outdated code can perfectly run without causing issues.

However, if this controller is present, all you need to do is to shutdown the servers at approx. the same time (which you need to do anyway, in the event of a major DB migration). Without a controller, you need to do that **and** bother a maintainer to update the DB at the same time.

I'm not that familiar with tg GitHub-server interactions, but you could always send a signal to server admins, or even the server itself, when a PR which requires immediate server reboot is merged.

Jordie0608 commented:
Red label stops our bot from marking things stale.

Jordie0608 commented:
I have two curiosities about how vg has been using this:
How does vg handle changes requiring more than one query, just multiple migrations in one PR?

And for changes that are too complicated does pomf just do those manually?
I couldn't see anything of the sort on your repo so I assume either you don't update the repo's schema file when that happens or it hasn't come up yet.

ShiftyRail commented:
> How does vg handle changes requiring more than one query, just multiple migrations in one PR?

You can either do that, or just make a more complicated migration datum. The migration datum actual "content" is its `up()` proc. The migration's `execute()` proc is a helper/wrapper. It takes raw SQL and executes it on your DB system, and logs an error if it fails to do so.

You can see in the BSQL PR that I have decided to keep the migration datums fairly simple and "atomic", but nothing stops you from running complete SQL code within a single migration.
It's just that most of our migrations have been small things.

>And for changes that are too complicated does pomf just do those manually?

Yeah. Although it is rare.
One instance I can think of is that we had, for the longest time, two concurrent DBs, one for everything, one of our library. When porting BSQL, it didn't work with that, so pomf just copied the library table in the DB we had.

You are correct that we do not always think of updating our `.sql` schema. It's actually something we should put in our contributing guidelines, so that our (eventual) downstreams can just run the .sql schema and have the migrations skipped.

Cyberboss commented:
![image](https://user-images.githubusercontent.com/8171642/83991245-2e583d00-a91a-11ea-92f7-bb81e42bcab6.png)
What did he mean by this?

optimumtact commented:
this needs more conversation with MSO, but is otherwise going nowhere

Cyberboss commented:
Personally, I'm of the opinion it should at least be included as a config option. Whether or not MSO wants to use it, I feel it'd be useful to have an authoritative migration system.

optimumtact commented:
I don't like dead code not running on tg cyberboss

Cyberboss commented:
Fair, though I find our migration tactics lacking

Cyberboss commented:
Looks like I stand alone here, too bad.
